### ISSUE TITLE
[HOLD for payment 2022-05-20] [$2000] When copy&paste multiple messages line breaks aren't maintained 

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7352]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app 
2. Copy a message with multiple line breaks 
3. Paste into another chat  

Steps 2
1. Send 3 separate messages
2. Copy all 3 messages
3. Paste in compose box

## Expected Result:
Line breaks should be maintained. The three messages would show on three lines

## Actual Result:
Line breaks are not maintained. The three messages are on one line.

## Workaround:
Unknown. 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.30-3
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
Related to https://github.com/Expensify/App/issues/5142  
![image - 2022-01-20T234205 354](https://user-images.githubusercontent.com/43995119/150467164-1825cd07-3ac3-482e-8652-ae1552f1f326.png)

**Expensify/Expensify Issue URL:**
**Issue reported by:**  @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1642095778439500

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @roryabraham (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mallenexpensify** [comment]
cc @kursat since you worked on https://github.com/Expensify/App/issues/5142
In case you wanted to work on this

**Comment by thienlnam** [comment]
@kursat Hey Kursat! Just looping back to this issue https://github.com/Expensify/App/pull/6732, could you comment some additional context where data-testID comes from and what significance it serves?

**Comment by roryabraham** [comment]
> What other purpose does data-testId serve? Because this selectionScrapper functionality seems to have been implemented recently, and I can't find the occurrence of this attribute elsewhere.

I too am trying to figure out where this comes from. It seems like it might be an implementation detail from either react-native-render-html or react-native-web 🤔

If that's the case, I think we need to decouple the `SelectionScraper` implementation from the `data-testid` attribute and instead make it reliant upon something we set explicitly in our app. I am still trying to understand all the implementation details of the `SelectionScraper` library, but it seems like at a high level it:

1. Uses the [Selection Web API](https://developer.mozilla.org/en-US/docs/Web/API/Selection) and then, for each [Range](https://developer.mozilla.org/en-US/docs/Web/API/Range) in the selection, finds the closest `div` tag with the `data-testid` attribute. Aggregate all those results under a single div.
    - This part seems like it's dependent upon low-level implementation details of react-native-web and is very unclear to me.
1. Takes the result of the previous step and parses it using [htmlparser2](https://github.com/fb55/htmlparser2)
1. Takes the result of the previous step, and recursively (for the dom tree parsed in the previous step) transform any `div` tags with the `data-testid` attribute to the actual tag specified in the attribute.
    - There is a comment [here](https://github.com/Expensify/App/blob/95c6ceb055b0c8c6a0bb15c5eb14561048bea426/src/libs/SelectionScraper/index.js#L23-L26) that alludes to that, but unless I'm mistaken this transform actually happens in a separate function [here](https://github.com/Expensify/App/blob/95c6ceb055b0c8c6a0bb15c5eb14561048bea426/src/libs/SelectionScraper/index.js#L95).
    - After each time we include a custom `<comment>` tag, we insert a `<br>` [here](https://github.com/Expensify/App/blob/95c6ceb055b0c8c6a0bb15c5eb14561048bea426/src/libs/SelectionScraper/index.js#L100).
1. "Re-render" the transformed html tree using [dom-serializer](https://www.npmjs.com/package/dom-serializer) [here](https://github.com/Expensify/App/blob/95c6ceb055b0c8c6a0bb15c5eb14561048bea426/src/libs/SelectionScraper/index.js#L131)
    - This doesn't actually render anything, but instead just simulates a render and returns the resultant html as a string.
1. Parse the html from the previous step to mark down using our own `ExpensiMark` library.

I have several thoughts about how this could all be simplified/cleaned up, but first I want to confirm that my high-level understanding of this is correct. @kursat Any comments?

**Comment by kadiealexander** [comment]
Posted to upwork: https://www.upwork.com/jobs/~0104b0a207be0dbbcb

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @rushatgabhane (`Exported`)

**Comment by MelvinBot** [comment]
Current assignee @roryabraham is eligible for the Exported assigner, not assigning anyone new.

**Comment by rushatgabhane** [comment]
> We can enclose the decoded HTML for the single line messages in a div with the data-testid passed explicitly as comment. The id gets attached to the DOM element and parsing is done accordingly.

Hmm I think that'll lead to an extra new line at the end of copied text.
Copying a single line of text will have a new line, which isn't ideal.

**Comment by kursat** [comment]
Hey @thienlnam, data-testid is added by react-native-render-html. 

@roryabraham all steps are correct.

@sig5's proposal looks like a workaround to me. So I am trying to find a better solution for this for a couple days. @roryabraham  If I can be able to find a better solution, we can also implement your improvement ideas.

**Comment by rushatgabhane** [comment]
>If I can be able to find a better solution, we can also implement your improvement ideas.

@kursat any updates on this one? Thanks!

**Comment by kursat** [comment]
Hey @rushatgabhane, I couldn't find enough time to find a proper solution yet. Please let me work some more this weekend. Hopefully, I will send a proposal on Monday.

**Comment by mallenexpensify** [comment]
@roryabraham should this job be doubled? I know @kursat said they're working on it but that shouldn't be a reason to not double because anyone can propose a solution and be hired

**Comment by roryabraham** [comment]
Yep, let's double this one.

**Comment by mallenexpensify** [comment]
Doubled price to $500 - https://www.upwork.com/jobs/~0104b0a207be0dbbcb

**Comment by roryabraham** [comment]
No new proposals here?

**Comment by kadiealexander** [comment]
Doubled price to $1000 - https://www.upwork.com/jobs/~0104b0a207be0dbbcb

**Comment by rushatgabhane** [comment]
> No new proposals here?

Nope.

@kursat please let us know if you have any updates, thanks ^^

**Comment by JalalMitali** [comment]
hello, I was looking in the code trying to fix it but I could not find where the chat room fragment is located I found "copytexttoclipboard.js" but it seems like it had nothing to do with this am I right please correct me thank you!

**Comment by ahmdshrif** [comment]
it's related to https://github.com/Expensify/App/issues/7644#issuecomment-1046210899

@rushatgabhane  when we copy we parse the copy HTML to markdown so we again use this function 
https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L197

which have some issues so I want to add these lines to support the correct copy here. 

```diff
pre : inputString
+.replaceAll("<div><div>","").replaceAll("</div></div>","").replaceAll('</div><div>', '<br/>')
.replace('<br></br>', '<br/>').replace('<br><br/>', '<br/>')
.replace(SLACK_SPAN_NEW_LINE_TAG + SLACK_SPAN_NEW_LINE_TAG, '<br/><br/><br/>').replace(SLACK_SPAN_NEW_LINE_TAG, '<br/><br/>')
```


https://user-images.githubusercontent.com/21364901/154840470-eb1d97e8-0520-4afd-a747-ff31f76b7f49.mov

**Comment by JalalMitali** [comment]
hello me too I was checking #3983 (https://github.com/Expensify/App/issues/3983)
it has to do with  html to markdown parsing :)

**Comment by rushatgabhane** [comment]
@ahmdshrif since you're modifying the same rule in #7644, 
could you please post an updated proposal which combines both for better clarity?

**Comment by ahmdshrif** [comment]
@rushatgabhane  this is work for both. 

```diff
 pre: inputString => inputString
+.replaceAll("<div><div>","").replaceAll("</div></div>","").replaceAll('</div><div>', '<br/>').replaceAll('</td></tr>', '<br/>')
.replace('<br></br>', '<br/>').replace('<br><br/>', '<br/>')
.replace(SLACK_SPAN_NEW_LINE_TAG + SLACK_SPAN_NEW_LINE_TAG, '<br/><br/><br/>').replace(SLACK_SPAN_NEW_LINE_TAG, '<br/><br/>'),
````

**Comment by rushatgabhane** [comment]
@ahmdshrif **_visit this file_** and try copying these lines.
https://github.com/Expensify/App/blob/403fc28984e9d96f3127257a3dc3ac6e66a75cd6/src/components/TextInputFocusable/index.js#L14-L16

I'm getting 2 line breaks instead of 1.
```
    maxLines: PropTypes.number,
<-- line break
<-- line break
    /** The default value of the comment box */
    defaultValue: PropTypes.string,
```

**Comment by michaelhaxhiu** [comment]
👋  should the price of this job double again today @kadiealexander ?

**Comment by kadiealexander** [comment]
@michaelhaxhiu Our engineers have advised not to double the price yet in this instance. :)

**Comment by roryabraham** [comment]
@ahmdshrif it seems like you're pretty close to the core issue here, but we're just lacking a complete proposal including the context for your changes. Why would we have double div tags, and why would we remove those from the markdown?

Furthermore, it looks like your solution preserves only _one_ newline, but not messages that have multiple newlines in a row.

**Comment by JalalMitali** [comment]
I thought this one is solved may you clarify for me what is still needed please ?

**Comment by ahmdshrif** [comment]
my proposal handle both newline messages that have multiple newlines in a row. and. but I will do some changes that are will work better for this issue. I am still working on optimizing the code block issue. 


the reason here when we map HTML to markdown we assume the new line only happens if we use the `<br/>` tag ond some few tags. 

but on our case here the new line is HTML use `div` instead of br 
but we can't map any div to a new line . for example that's what we have when we copy the  4 message 


```
<div><div><div><div><div><div><div><div><div><div><div>line1</div></div></div></div></div><div></div></div></div></div></div></div><div><div><div><div><div><div><div><div><div><div>line2</div></div></div></div></div><div></div></div></div></div></div></div><div><div><div><div><div><div><div><div><div><div>line3</div></div></div></div></div><div></div></div></div></div></div></div><div><div><div><div><div><div><div><div><div><div>line4</div></div></div></div></div></div></div></div></div></div></div>
```

lot of div and if we map the close div as new line will have many extralines between message . 

so we just need to map the dev that have text between open and close tags to new line . 

for example we just will map  
`<div>line1</div>`
`<div>line1</div>`

**Comment by ahmdshrif** [comment]
my proposal handle both newline messages that have multiple newlines in a row. and. but I will do some changes that are will work better for this issue. I am still working on optimizing the code block issue. 


the reason here when we map HTML to markdown we assume the new line only happens if we use the `<br/>` tag ond some few tags. 

but on our case here the new line is HTML use `div` instead of br 
but we can't map any div to a new line . for example that's what we have when we copy the  4 message 


```
<div><div><div><div><div><div><div><div><div><div><div>line1</div></div></div></div></div><div></div></div></div></div></div></div><div><div><div><div><div><div><div><div><div><div>line2</div></div></div></div></div><div></div></div></div></div></div></div><div><div><div><div><div><div><div><div><div><div>line3</div></div></div></div></div><div></div></div></div></div></div></div><div><div><div><div><div><div><div><div><div><div>line4</div></div></div></div></div></div></div></div></div></div></div>
```

lot of div and if we map the close div as new line will have many extralines between message . 

so we just need to map the dev that have text between open and close tags to new line . 

for example we just will map  
`<div>line1</div>`
`<div>line2</div>`
`<div>line3</div>`

also will not map 
`<div>line4</div>`
to new line because this will give extra line at the end . 

also will ignore all other div tags 

my suggestion here is to delete any empty div inputString.replaceAll('<div></div>', '')
and then map only the div that have open div next not it to new line . 
.replaceAll('</div><div>','<br/>')

I test this on many cases and work fine  

so the final change here will be 

```diff
 pre: inputString => inputString
+.replaceAll("<div><div>","").replaceAll("</div></div>","").replaceAll('</div><div>', '<br/>').replaceAll('</td></tr>', '<br/>')
.replace('<br></br>', '<br/>').replace('<br><br/>', '<br/>')
.replace(SLACK_SPAN_NEW_LINE_TAG + SLACK_SPAN_NEW_LINE_TAG, '<br/><br/><br/>').replace(SLACK_SPAN_NEW_LINE_TAG, '<br/><br/>'),
````

**Comment by ahmdshrif** [comment]
just work on this to optimize it with regex 

this part is what we need only to add to fix this one 
`.replaceAll(/<\/div>(<[^/])/g, "<br/>$1")`
this regex map any close div that is not in the end to `br`
and that is the same as what I do before but this is perfect and handles the case on copy code from VSCode. 

and that's the final proposal that fixes both issues as you request  @rushatgabhane  will be this: - 


```diff 
 pre: inputString => inputString
+.replaceAll(/<\/div>(<[^/])/g, `<br/>$1`).replaceAll(/(<tr.*?<\/tr>)/g,`$1<br/>`).replaceAll('<br/></tbody>',"")
.replace('<br></br>', '<br/>').replace('<br><br/>', '<br/>')
.replace(SLACK_SPAN_NEW_LINE_TAG + SLACK_SPAN_NEW_LINE_TAG, '<br/><br/><br/>').replace(SLACK_SPAN_NEW_LINE_TAG, '<br/><br/>'),
```

I will explain more about the second regex on the second issue. since it's more related to copying code from Github. 



https://user-images.githubusercontent.com/21364901/155846490-2f432e9d-5aba-4e49-b05e-fa9c7a70c92b.mov

**Comment by blacknight3** [comment]
**Proposal**<hr>
let's explain what is the issue and  how  these three issues are related to each other https://github.com/Expensify/App/issues/7352 ,  https://github.com/Expensify/App/issues/7912 , https://github.com/Expensify/App/issues/7644

when we try to copy some text from Expensify what happens is this text is converted to a bad HTML code then `htmlToMarkdown` function tries to convert this bad HTML to text and put it into our clipboard so we can paste it anywhere, the problem is `htmlToMarkdown` function can't convert this bad HTML to text correctly so when we try to paste it into Expensify or into any text editor we get the text in a wrong format.

here is what happens when we try to copy `1234` from Expensify to a text editor we get `4321`.

![A](https://user-images.githubusercontent.com/100511864/156046757-80d03974-4e47-451b-8689-322eac3834e3.png)

fixing the issue by trying to convert this bad HTML to text while pasting it into Expensify isn't a good solution, it works for a simple and small HTML, not for a complex one and the problem exists when we try to paste it into any text editor,  so we won't do anything to `handlePaste` and won't handle any HTML there, it works fine by default, instead we will correct our HTML while coping so we can paste it anywhere in a correct format.

converting HTML to text is not easy, and one line regex can't do it correctly, we don't only remove HTML elements because some divs maybe represent a new line and others represent something else.

`html-to-text` module is a good solution, here is their repo  https://github.com/html-to-text/node-html-to-text

Unfortunately, `html-to-text` can't convert this bad HTML correctly but it is converting better than our function `htmlToMarkdown`

so we will fix some issues in this bad HTML manually then pass the output to `html-to-text` and we will get the correct text.

here is how it work 


https://user-images.githubusercontent.com/100511864/156055569-3ade56c9-aa84-43d1-97f1-30c6fdccf99d.mov

**Comment by K4tsuki** [comment]
#### proposal

The problem is the converter couldn't detect line break across `ReportActionItem` we could assist this by adding:
`<br />` in :

 Above:

https://github.com/Expensify/App/blob/d6e64b37e2cc38a692b0778de3348e6e7da924f0/src/pages/home/report/ReportActionItem.js#L159

And Below:
https://github.com/Expensify/App/blob/d6e64b37e2cc38a692b0778de3348e6e7da924f0/src/pages/home/report/ReportActionItemSingle.js#L89

**Comment by kadiealexander** [comment]
@rushatgabhane what do you think of the above proposals? 😊

**Comment by rushatgabhane** [comment]
For this [proposal](https://github.com/Expensify/App/issues/7352#issuecomment-1052139980) @ahmdshrif 

Try copy paste for these lines. Something to do with the text made bold using `*` 
```
*Lorem Ipsum* is simply dummy text of the printing and typesetting industry.

Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.

Hello 👋
```

https://user-images.githubusercontent.com/29673073/157110753-1dc2b65e-8802-4a32-8b1b-2c7d73391055.mov

**Comment by rushatgabhane** [comment]
@blacknight3 it'd be great if we can find a solution without using any extra libraries

**Comment by rushatgabhane** [comment]
@K4tsuki I think that's same as https://github.com/Expensify/App/issues/7352#issuecomment-1020288547 which seems like a hack

**Comment by kadiealexander** [comment]
Doubled price to $2000 - https://www.upwork.com/jobs/~0104b0a207be0dbbcb

**Comment by blacknight3** [comment]
### proposal

this solution will fix https://github.com/Expensify/App/issues/7352 ,https://github.com/Expensify/App/issues/7644 but not https://github.com/Expensify/App/issues/7912  without using any  packages

- In src/components/TextInputFocusable/index.js comment this if statement `if (types.includes(TEXT_HTML)) {...}` 
- in src/libs/SelectioinScraper/index.js add this function : 
```
const fixReverseHTML = (reversedHtml) => {
    let zindexes = reversedHtml.split('<div class="css-view-1dbjc4n r-flexDirection-1d5kdc7 r-transform-1qc3rpd" ')
    zindexes.shift()
    if (zindexes.length == 0) {
        return reversedHtml
    }
    for (let i = 0; i < zindexes.length; i++) {
        if (zindexes[i].includes('<div dir="auto" class="css-text-901oao r-userSelect-1xnzce8" style="color: rgb(11, 27, 52); font-family: GTAmericaExp-Regular; font-size: 15px; line-height: 20px; text-align: left;"></div>') || zindexes[i].includes('<div dir="auto" class="css-text-901oao r-userSelect-1xnzce8" style="color: rgb(11, 27, 52); font-family: GTAmericaExp-Regular; font-size: 15px;"></div>') || zindexes[i].includes('css-view-1dbjc4n r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010')) {
            zindexes.splice(i, 1)
        }
    }
    if (zindexes.length == 1 && zindexes[0].includes('<div class="css-view-1dbjc4n r-position-u8s1d" style="right: 16px; top: -32px; z-index: 1;">')) {
        zindexes[0] = `<div class="css-view-1dbjc4n r-flexDirection-1d5kdc7 r-transform-1qc3rpd" ` + zindexes[0].replace('<div class="css-view-1dbjc4n r-position-u8s1d" style="right: 16px; top: -32px; z-index: 1;">', '<div>')
        return zindexes
    }
    for (let j = 0; j < zindexes.length - 1; j++) {
        if (zindexes.length > 1) {
            for (let i = 0; i < zindexes.length; i++) {
                if (i + 1 < zindexes.length) {
                    let currenZindexValue = parseInt(zindexes[i].split(';">')[0].slice(16))
                    let nextZindexValue = parseInt(zindexes[i + 1].split(';">')[0].slice(16))
                    if (currenZindexValue > nextZindexValue) {
                        let currentIndex = zindexes[i]
                        zindexes[i] = zindexes[i + 1]
                        zindexes[i + 1] = currentIndex
                    }
                    else {
                    }
                }
                else {
                }
            }
        }
    }
    for (let i = 0; i < zindexes.length; i++) {
        zindexes[i] = `<div class="css-view-1dbjc4n r-flexDirection-1d5kdc7 r-transform-1qc3rpd" ` + zindexes[i]
    }
    if (!zindexes[0].includes('<div class="css-view-1dbjc4n r-position-u8s1d" style="right: 16px; top: -32px; z-index: 1;">')) {
        zindexes[0] += '<br>'
    }
    if (zindexes[zindexes.length - 1].includes('<div class="css-view-1dbjc4n r-position-u8s1d" style="right: 16px; top: -32px; z-index: 1;"></div>')) {
        zindexes[zindexes.length - 1] = zindexes[zindexes.length - 1].replace('<div class="css-view-1dbjc4n r-position-u8s1d" style="right: 16px; top: -32px; z-index: 1;">', '<div>')
    }
    for (let i = 0; i < zindexes.length; i++) {
        if (i < zindexes.length - 1) {
            zindexes[i] = zindexes[i].replaceAll(/<div class="css-view-1dbjc4n r-position-u8s1d" [^>]*>?<\/div>/gm, '<br>')
        }
    }
    let correctedHtml = zindexes.join('')
    return correctedHtml
}
```

- call it here `const domRepresentation = parseDocument(fixReversedHTML(selectionHtml));`

Result:

https://user-images.githubusercontent.com/100511864/157664522-4891c7aa-f582-42fa-81f3-00e205064996.mp4

**Comment by ahmdshrif** [comment]
@rushatgabhane  these changes fix all cases. 

https://github.com/ahmdshrif/expensify-common/commit/c99333c617c4833f4b7609fb2c50799784f2fd85


https://user-images.githubusercontent.com/21364901/157667557-d121a22b-ad50-4bec-8766-91d52f4a544a.mov

**Comment by rushatgabhane** [comment]
@ahmdshrif could you please help me understand your proposal? 
Like what was the cause, and what fixed it.

**Comment by rushatgabhane** [comment]
@blacknight3 sorry, I'm not able to understand what `fixReverseHTML()` does. Could you please give me a high level overview of what's happening?

Anyway, we already have [ExpensiMark](https://github.com/ahmdshrif/expensify-common/blob/main/lib/ExpensiMark.js), so why not use it instead?

**Comment by ahmdshrif** [comment]
@rushatgabhane  the regex fails because it's not the true way to handle this I can do some changes to handle the case of blood text but expect to fail on other cases we do not try it. 

the problem is we do not handle the case of mapping block. 
by block, I mean div for example. since it can be used as a new line comment also can be used as block .

I do many tests on when the div can be a newline.  I have 4 examples will explain it 

1-if we have div inside the text  `<div> line 1 <div>line 2</div></div>` we have new line 
<div> line 1 <div>line 2</div><div/>

2- if open and close div inside the text `<div>line 1 <div></div>line 2 <div>` we have new line 
<div>line 1 <div></div>line 2 <div>

3- if many div then word  `<div><div></div>line 1</div>` we  not have new line 
<div><div></div>line 1</div>

4- if div then word in div the word then div  new line `<div><div>line 1</div>line 2</div>`
<div><div>line 1</div>line 2</div>

5- this all will be newline `<div>line 1<div>line 2</div>line 3</div>`
<div>line 1<div>line 2</div>line 3</div>


the conclusion here we need to check the text  (NOT EMPTY OR WHITE SPACE )if it has an open or closed dive after it  we will add a new line 
but we need also to ignore all other HTML tags since we can't count them as text 

so the solution here start after applying all rules by mapping any open or close div or comment  to [block] to not split with `stripHTML` function then split the HTML  

```ts
generatedMarkdown = generatedMarkdown.replaceAll(/<div.*?>|<\/div>|<comment.*?>|<\/comment>/gm,"[block]")
return Str.htmlDecode(this.mapBlock(Str.stripHTML(generatedMarkdown)));
 ```
 
 then we send it to mapBlock function 
 
 in mapblock i split the text by "[block] "t o an array with text between blocks 

 ```ts
      const splitText = htmlString.split("[block]")
```
then check if this text is not empty spaces by trim()

if it's not empty we can add a new line to it. 



this is a complete function. 

```ts
mapBlock (htmlString){
        const splitText = htmlString.split("[block]");
        let joinedText = splitText[0]
        for (let index = 1; index < splitText.length -1; index++) {
            if (splitText[index].trim()) {
                joinedText += splitText[index]+ '\n'
            }
        }
        return joinedText
    }
```

**Comment by blacknight3** [comment]
@rushatgabhane  `fixReverseHTML()` function does the following : 

**1. Arrange HTML elements z-index ascending**
when I try to copy these lines `line1 line2 line3` I get it in a reversed order `line 3line 2line 1`, so first I have to arrange HTML elements z-index ascending to get it in the correct order. I use Expensify version 1.1.30-3

https://user-images.githubusercontent.com/100511864/157856682-8243be3a-be52-4513-b759-e79b1adda224.mov


2. **Replacing some `<div>` into `<br>`**
In  HTML code this `<div class="css-view-1dbjc4n r-position-u8s1d" style="right: 16px; top: -32px; z-index: 1;"><\div>` represent a line break, so we need to replace it with `<br>`, **note** : our converter convert it to `<div></div>` and it is the problem.

3.  **Removing extra line breaks**
after arranging HTML elements ascending by z-index I need to remove the line break from the first element and add it to the last one,
remove extra line break from the middle elements.

### Note:
if you don't have the issue in step 1 you don't need `fixReverseHTML()` function, only do the following : 
in src/libs/SelectioonScraper/index.js add this if statement and you will get the same result.
```
+   if(dom.attribs && dom.attribs["class"]){
+       if(dom.attribs['class']== "css-view-1dbjc4n r-position-u8s1d"){
+           dom.children.push(new Element('br', {}));
+       }
+   }
    // We need to preserve href attribute in order to copy links.
    if (dom.attribs && dom.attribs.href) {
        domAttribs.href = dom.attribs.href;
    }
``` 
### Finally 
we still can't copy any HTML element because what our converter does  'if I'm not mistaken' is take HTML code and convert most of the HTML elements into `<div></div>` then extract text between these divs.

ex: `hello <h1>how </h1><div>are</div> you ` ---> our converter ---> `hello how are you`

**Comment by K4tsuki** [comment]
@rushatgabhane I thought we have library with html styling awareness to convert html with styling to markdown, but actually we make our own converter. At this rate if something changes in `ReportActionItem`or component inside it then we must modify the converter. So, it will not be one time solution to fix current and future change of `ReportActionItem`.

**Comment by rushatgabhane** [comment]
@ahmdshrif thanks!
From what I understand,  you're adding `[block]` to keep track of indexes to add `'\n'`, because you'll lose them after `Str.stripHTML()`.

But.. `Str.stripHTML()` doesn't remove `'\n'`. So why not add it instead of `[block]` in the first place?

**Comment by ahmdshrif** [comment]
Because not all blocks will mapped to '/n' 
The condition here is we have word between blocks (word not spacess or html tag)

**Comment by rushatgabhane** [comment]
@blacknight3 thanks for explaining your proposal, but it breaks when copying text from top to bottom 

https://user-images.githubusercontent.com/29673073/158485560-9d5a7e36-0269-4c56-bf66-758de2351808.mov

**Comment by rushatgabhane** [comment]
@ahmdshrif we have an extra line breaker when copying italic text

https://user-images.githubusercontent.com/29673073/158487665-e480fb9d-4c0e-4001-8797-3f9a458e3f32.mov

**Comment by blacknight3** [comment]
@rushatgabhane it seems a browser issue because this code works fine on chrome [linux] but doesn't work on firefox [linux].

### updated proposal
in src/libs/SelectioonScraper/index.js 
```
+   if(dom.attribs && dom.attribs["class"]){
+       if(dom.attribs['class']== "css-text-901oao r-userSelect-1xnzce8" && !dom.attribs['data-testid'] && !dom.parentNode.attribs['data-testid']){
+           dom.children.push(new Element('br', {}));
+       }
+   }
    // We need to preserve href attribute in order to copy links.
    if (dom.attribs && dom.attribs.href) {
        domAttribs.href = dom.attribs.href;
    }
```
this code works on both firefox [linux] and chrome [linux], also it can skip the italic issue

https://user-images.githubusercontent.com/100511864/158608930-4d4b79df-f100-4a0f-a0fb-e7238684ad41.mp4

.

**Comment by rushatgabhane** [comment]
@ahmdshrif it's not copying line breakers from vscode, are you able to fix that?

**Comment by rushatgabhane** [comment]
@blacknight3 sorry but your proposal seems like a hack, and It's very similiar to https://github.com/Expensify/App/issues/7352#issuecomment-1020288547
Please correct me if I'm wrong

**Comment by ahmdshrif** [comment]
@rushatgabhane  yes get it the trim function skip the new line. 
so it fixed by this line `            if (splitText[index].trim()||splitText[index].match(/\n/)) {        `

```ts
for (let index = 1; index < splitText.length -1 ; index++) {
            if (splitText[index].trim()||splitText[index].match(/\n/)) {        
                if (splitText[index].match(/\n$/)) {
                    joinedText += splitText[index]
                } else {
                    joinedText += splitText[index] + '\n'
                }
            }
        }
```

I also note we have a new line at the end also add some code to fix it 

check the final version here https://github.com/ahmdshrif/expensify-common/blob/main/lib/ExpensiMark.js

I will clean up the code more when I get hired but now focus on cover all cases for both issues. 

https://user-images.githubusercontent.com/21364901/158702115-72fbc476-5cf4-40e8-9e22-bb74f3dd1093.mov

**Comment by rushatgabhane** [comment]
Okay sounds good. Here are some edge cases.

https://user-images.githubusercontent.com/29673073/158704476-9bb0a373-d0dd-42b8-a6f3-d836f9a9056e.mov

@roryabraham I'd love your opinion on @ahmdshrif's approach. It seems like we're gonna have too many cases to test, and debuging the code later on won't be easy. Here's the diff for you - https://github.com/Expensify/expensify-common/compare/main...ahmdshrif:main

I honestly wouldn't mind using some lib to handle things for us

**Comment by ahmdshrif** [comment]
I believe the hard task is done now and we have a few steps to finish. 
we will add some test cases related to all tags that we map 

but for the case you mention here, it's already happened on prod and that's something that needs to fix from the `modifyTextForQuote`  first. 

https://user-images.githubusercontent.com/21364901/158779904-3158688c-fa9c-469a-ada2-095cfc3541e7.mov

I can work on it but still waiting for your decision.

**Comment by rushatgabhane** [comment]
Thanks @ahmdshrif 

C+ reviewed 🎀 👀 🎀

@roryabraham I think we should hire @ahmdshrif, we can then work together on all the test cases.
@ahmdshrif's [proposal](https://github.com/Expensify/App/issues/7352#issuecomment-1064853399) looks good to me.

**Comment by roryabraham** [comment]
Okay. Honestly, I had some difficulty following the language used throughout this discussion, but the videos give me enough confidence to think that we're on the right track.

> @roryabraham We might be compounding on a wrong path, so I'd love your opinion on @ahmdshrif's approach. It seems like we're gonna have too many cases to test, and debuging the code later on won't be easy.

I agree that there's a substantial burden on maintenance in our Markdown parser, but that predates my time at Expensify so I don't have all the context around why we decided to build our own Markdown parser. Generally I think the reason is that we wanted to have full control over the markdown syntax we choose to support not to support in our chat application.

**Comment by MelvinBot** [comment]
📣 @ahmdshrif You have been assigned to this job by @roryabraham!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by kadiealexander** [comment]
Have sent contracts to @ahmdshrif and @rushatgabhane to ensure they're paid when this is completed. Please accept when you can!

**Comment by ahmdshrif** [comment]
pr here @rushatgabhane  @roryabraham 
https://github.com/Expensify/expensify-common/pull/436

**Comment by roryabraham** [comment]
Looks like the PR is almost ready. @rushatgabhane mentioned that he wants to review, so I'm waiting on the C+ 👍  before I do a final review.

**Comment by roryabraham** [comment]
PR review is still ongoing

**Comment by roryabraham** [comment]
[Expensify-common PR](https://github.com/Expensify/expensify-common/pull/436) was merged just a few minutes ago. Next up is the NewDot PR.

**Comment by mallenexpensify** [comment]
@roryabraham in case anything is _weird_ with this issue, can you drop a comment in denoting payment date if one isn't auto-added?

**Comment by kadiealexander** [comment]
See update above! Begone, Melvin.

**Comment by kadiealexander** [comment]
@roryabraham do we need to issue payment for anything so far? I'm unclear on if this issue has anything unusual I need to manage.

**Comment by rushatgabhane** [comment]
@kadiealexander this issue should be setttled on May 20.
PR https://github.com/Expensify/App/pull/8743 was deployed to production on May 13.

**Comment by kadiealexander** [comment]
Thanks @rushatgabhane!

**Comment by kadiealexander** [comment]
Paid!!! Thanks for your work team.

**Comment by rushatgabhane** [comment]
Cheers!
